### PR TITLE
Added config_flow, hacs and switch.

### PR DESCRIPTION
**Pull request by RenierM26**
I'm busy with a pull request on hassio to update official component. Perhaps the community can benefit from the changes while we wait for a review? The structure allows for easy adding of platforms (ie sensor platform for battery status and light level)   

-Added HACS file for easy updating.
-Added config flow. (Tests the MAC address is valid via pygatt - hassio git ci can't pass with pyblue)
-Added hassio switch component with a device descriptions.
-Small changes to recently merged cover platform pull request. (basically depreciated Dict typing fix, config flow and device identification - requirement to match sensor entities with device for future contribution) 

Breaking change

-Curtain entries will need to be reconfigured via the integrations gui and removed from yaml config.
-Will import and create switchbot bot entites.

### CODE REVIEW

**Comment by RenierM26** [comment]
*Breaking change*

-Curtain entries will need to be reconfigured via the integrations gui and removed from yaml config.
-Will import and create switchbot bot entites.

File: custom_components/switchbot-curtain/config_flow.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    try:
+        adapter.start()
+        device = adapter.connect(mac, address_type=pygatt.BLEAddressType.random)
+    finally:
+        adapter.stop()
+
+    return device
+
+
```
Comment (RenierM26) [inline custom_components/switchbot-curtain/config_flow.py]: Found a potential problem here. Please wait while I fix the device connection test.

Comment (RenierM26) [reply custom_components/switchbot-curtain/config_flow.py]: -Fixed invalid command error with adapter.start(reset_on_start=False)
-Added thread lock as multiple connections could cause issues with pygatt.

File: custom_components/switchbot-curtain/cover.py [hunk 2]
```diff
@@ -160,7 +157,18 @@ def current_cover_position(self):
...
+            "name": self._name,
+            "model": "Curtain",
+            "manufacturer": MANUFACTURER,
+        }
+
+#    def update(self):
+#        """Update device attributes."""
+#        self._device.update()
+#        self._light = self._device.get_light_level()
+#        self._battery = self._device.get_battery_percent()
```
Comment (RenierM26) [inline custom_components/switchbot-curtain/cover.py]: Found a problem with this code. Could be caused by multiple connection attempts locking the bluetooth adapter.

Working on fix.

Comment (RenierM26) [reply custom_components/switchbot-curtain/cover.py]: Fix issue. It now works well with multiple devices. ;-)

**Comment by RenierM26** [comment]
@alextud,

Tested pull request with multiple devices and works well.

Thanks.

**Comment by RenierM26** [comment]
Should also resolve #12

**Comment by RenierM26** [comment]
https://github.com/alextud/ha-switchbot-curtain/blob/7c41a17ca17a8b3ae53def6a4c6ab7e639a63f42/custom_components/switchbot-curtain/cover.py#L165

This line results in state history being spammed with irrelevant updates (Every update results in state change, even if it hasn't). Busy fixing.

**Comment by alextud** [comment]
Hey @RenierM26, I have a flaky connection to my switchbot (not always is working - assuming overcrowded network )

Is it possible to increase polling time (now - 60 seconds) to a custom one, through yaml or web interface ? My self I don't use this information - as the curtains is used only 2 a day (mostly trough scene automation).

**Comment by alextud** [comment]
I've seen that you also integrated original switchbot implementation in this repo, is the original developer ok with this ?

**Comment by RenierM26** [comment]
> I've seen that you also integrated original switchbot implementation in this repo, is the original developer ok with this ?

I haven't asked him but I see you have credited him in the manifest file ? @danielhiversen 

I have also changed approx +-42% of the code so I don't think it will be a problem.

**Comment by RenierM26** [comment]
> Hey @RenierM26, I have a flaky connection to my switchbot (not always is working - assuming overcrowded network )
> 
> Is it possible to increase polling time (now - 60 seconds) to a custom one, through yaml or web interface ? My self I don't use this information - as the curtains is used only 2 a day (mostly trough scene automation).

Good idea. I can add options to the config flow for this.

**Comment by Danielhiversen** [comment]
It is ok with me.
Hope you will consider adding this to the main HA repo also

**Comment by RenierM26** [comment]
Added config flow options for the various switchbot api timeouts.

Local test pass.

**Comment by RenierM26** [comment]
Hi @alextud,

I'm done. Any other suggestions?

File: custom_components/switchbot-curtain/const.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+DOMAIN = "switchbot-curtain"
+MANUFACTURER = "switchbot"
+
+ATTR_CURTAIN = "curtain"
+ATTR_BOT = "bot"
+DEFAULT_NAME = "Switchbot"
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_RETRY_TIMEOUT = 1
+DEFAULT_TIME_BETWEEN_UPDATE_COMMAND = 10
```
Comment (alextud) [inline custom_components/switchbot-curtain/const.py]: @RenierM26 isn't the DEFAULT_TIME_BETWEEN_UPDATE_COMMAND to fast = 10 seconds. I think that we can keep default to 5 minutes ?

Comment (alextud) [reply custom_components/switchbot-curtain/const.py]: or 10 = 10 minutes ... ?
Screenshot 2021-05-27 at 18 42 15

Comment (alextud) [reply custom_components/switchbot-curtain/const.py]: After looking into pySwitchbot code, time between updates = waiting time to get newer information after a command was issued to the bot. (I think that we need to wait let's say 10 seconds after bot moved to correct position and a scanned is performed)

https://github.com/Danielhiversen/pySwitchbot/blob/1357c7c54403296172ad51fa3db2b3748b0e56f5/switchbot/__init__.py#L124

In order to control polling time, we need to update SCAN_INTERVAL which is now to 35 seconds ...

So I think that "time between updates" is misleading in this context ...

Comment (RenierM26) [reply custom_components/switchbot-curtain/const.py]: Hi @alextud,

I've also noticed that. I'm busy looking into a wat to set SCAN_INTERVAL the same as DEFAULT_TIME_BETWEEN_UPDATE_COMMAND. The hassio update and async_update functions have a hard coded warning if the update takes longer than 10 seconds. (in hassio entity helpers code)

I'm thinking of changing polling to False and scheduling ha updates based on DEFAULT_TIME_BETWEEN_UPDATE_COMMAND.

Comment (RenierM26) [reply custom_components/switchbot-curtain/const.py]: https://github.com/home-assistant/core/blob/3ca7eb9440f8b6b4b4825b8b550903a699b3de03/homeassistant/helpers/entity.py#L527

Comment (RenierM26) [reply custom_components/switchbot-curtain/const.py]: You are correct about "time between updates".  Will need to factor this in as well.

**Comment by RenierM26** [comment]
Hi @alextud,

I have submitted a pull request on pySwitchbot. It should fix a few issues. (Battery level incorrect for curtains, improvement for update methods and battery status and switch status for bot)

Once merged, I will update this integration.

**Comment by RenierM26** [comment]
Due to btle device connection limits I suspect it will be better to handle updates via a coordinator.  I'm going to try that next.

The integration works but it spams the logs when it can't connect. (BTLE spec allows for only 1 connection at a time)

**Comment by RenierM26** [comment]
Hi @alextud,

It now uses an update coordinator and should be much more reliable.  My pull request on the API is still being reviewed so this is kind of a proof of concept until the pull request is reviewed and merged. (Manifest file points to my pull branch on git instead of pypi)

**Comment by RenierM26** [comment]
@alextud,

Updates are working properly (with multiple devices ;-). Next is to change the command execution to a coordinator based model as well.

**Comment by 88fingerslukee** [comment]
@RenierM26 I'm so stoked on your work. Thank you so much! When will we be able to test these updates out?

**Comment by RenierM26** [comment]
Hi @88fingerslukee and @alextud ,

I done with this pull request (as of now). Just waiting for Danielhiversen to review and commit the upstream api changes. 

Your welcome to test. Just bare in mind that an API change could still happen that requires a change or two.

**Comment by RenierM26** [comment]
Hi @alextud,

The upstream API changes have been merged. This pull request is ready for review.

**Comment by 88fingerslukee** [comment]
I'm very excited for this...

**Comment by 88fingerslukee** [comment]
@RenierM26 I'm getting a gatttool.py timeout error when I test the files from your repo. It can't find connect to either of my curtains.  When I run gatttool interactively from the CLI I get a connection refused error.  bluetoothctl discovers and connects just fine, however.  Any thoughts on this?

My curtains were working under @alextud's old integration

UPDATE: Nevermind, my signal strength is not good (-92 and -88). I need to work on getting my antenna in a better position.

**Comment by 88fingerslukee** [comment]
FURTHER UPDATE

@RenierM26 
It seems gatttool is broken in Fedora 33 and deprecated in Fedora 34. My server is currently running Fedora 33. I've tried running gatttool on a second machine running Ubuntu 20.04 and it connects no problem. I've tried Fedora 33 AND 34 on that same second machine and it gives me the "connection refused" error.  gatttool is not installed by default on Fedora 34, one must install the bluez-deprecated package in order to use it. 

Two questions:

1.) Is gatttool the only way of making this work?
2.) will gatttool be deprecated in future versions of ubuntu or debian or whatever?

**Comment by 88fingerslukee** [comment]
@RenierM26 thanks for all the work here. I hope I'm not being a total downer.

I have tried your latest commit. I first got a btle.py Permission Denied error.  I fixed that by issuing:

`sudo setcap 'cap_net_raw,cap_net_admin+eip' btle.py`

But the config still fails without any error in the logs.  Is there some way of increasing the log level to figure out what's going on here?

**Comment by RenierM26** [comment]
Hi @88fingerslukee,

Would you be able to open python and run the following?

Import switchbot
switchbot.GetSwitchbotDevices().discover()

It should output a python dict with all the switchbots nearby. The config flow basically uses this to verify that the Mac is entered correctly. If there are any errors then we should be able to figure out the problem.

**Comment by 88fingerslukee** [comment]
`>>> Import switchbot`
`  File "<stdin>", line 1`
`    Import switchbot`
`           ^`
`SyntaxError: invalid syntax`


this is weird, no? what am I doing wrong?

**Comment by 88fingerslukee** [comment]
@RenierM26 
I tried this again this morning and it worked. The problem was the capital "I" in Import. Lowercase worked.

It definitely discovered the devices.  Now what?

UPDATE: I have the integration working now. I had to enter the MAC address in lowercase letters, I was entering it in uppercase.  
This is great!

**Comment by RenierM26** [comment]
Hi @88fingerslukee,

Thank your for the feedback. I'll try to add some code that converts the input to lowercase.

**Comment by 88fingerslukee** [comment]
@RenierM26 thank you so much for your work. If you can't add the code, we should at least make sure the instructions call it out.

**Comment by 88fingerslukee** [comment]
@alextud @RenierM26 do we know when these updates will be pushed to HACS?

**Comment by LouiS22** [comment]
Sorry to report this, but on my Intel NUC this version has problems (Error connecting to Switchbot device), and eventually the device become unavailable. The original version works flawlessly.

**Comment by 88fingerslukee** [comment]
@RenierM26 I'm also having a problem now. The integration seemed to be causing stability issues with HASS and it would shut down randomly. Then, one of my two curtain devices becomes unavailable. Now, this solves the stability issue but every time I remove and re-add the device the server starts shutting down. 

What kind of logs do you need for this?

**Comment by alextud** [comment]
When viewing integration configuration, last input description is not visible for me.
Screenshot 2021-07-25 at 13 20 07

**Comment by alextud** [comment]
@RenierM26 Switchbot in some cases reports that closed position is at 1% (when closing curtain). It seems that homekit is using this position to report if curtain is open/closed. Can we report 1 to be as 0 ex:

```
def current_cover_position(self) -> int:
         """Return the current position of cover shutter."""
         position = self.coordinator.data[self._idx]["data"]["position"]
         if position == 1: // homekit is using this to show if open / closed state..
              position = 0
         return position
```

Otherwise when I'm closing curtain from homekit is always reports as in progress ...

**Comment by alextud** [comment]
> @RenierM26 I'm also having a problem now. The integration seemed to be causing stability issues with HASS and it would shut down randomly. Then, one of my two curtain devices becomes unavailable. Now, this solves the stability issue but every time I remove and re-add the device the server starts shutting down.
> 
> What kind of logs do you need for this?

Can you increase time between updates to be more than 1 min, let's say 5 or 10 minutes...

**Comment by alextud** [comment]
After a successful comand HA logs a warning:


```
2021-07-25 14:36:50 DEBUG (SyncWorker_8) [switchbot] Sending command, 570F450105ff00
2021-07-25 14:36:50 INFO (SyncWorker_8) [switchbot] Successfully sent command to Switchbot (MAC: dc:6d:a0:60:c2:ab)
2021-07-25 14:36:50 DEBUG (SyncWorker_8) [switchbot] Disconnecting
RuntimeWarning:
    coroutine 'DataUpdateCoordinator.async_request_refresh' was never awaited
```

**Comment by 88fingerslukee** [comment]
@RenierM26 I'm testing these changes now. I've had the latest version installed since last night. So far no lockups on HA

**Comment by 88fingerslukee** [comment]
So far so good. No HA lockups and it seems to be more responsive.

**Comment by markusstephanides** [comment]
Hi I'm also testing @RenierM26s changes - working really good so far! I only wondered if you plan to integrate some sort of quiet mode since the quiet mode of the device is not really quiet. I thought of something like whenever you change the curtain position you can specify a duration that it should take. Then the integration would split the operation into little pieces to move the curtain step by step with little breaks in between to keep the noise as low as possible and the curtain would reach it's final position at the end of the specified duration time.

**Comment by markusstephanides** [comment]
I just had a hang up of home assistant after trying to move the curtains. Now they won't move again. Works instantly via the app though :/

**Comment by RenierM26** [comment]
Hi There,

Thank you for the help everyone! All of the code was merged into the official hass integration and should be available in the next hass release. ;-) (Config Flow, curtain support and the sensors)

**Comment by 88fingerslukee** [comment]
@RenierM26 Thank you for all your work! Where will we make comments on issues going forward? Also, will I have to delete my devices and re-create them when the official integration drops?

**Comment by RenierM26** [comment]
Hi @88fingerslukee,

The platform name on the hass integration is different so you'll need to re-create your devices. You can always open an issue on the  hass github repository if you have any problems or comments.

**Pull request closed without merging**